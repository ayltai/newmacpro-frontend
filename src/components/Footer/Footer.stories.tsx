import { Typography, } from '@mui/material';
import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';

import { default as FooterLink, } from '../FooterLink';
import { default as Footer, } from './Footer';

const meta : Meta<typeof Footer> = {
    component : Footer,
    tags      : [
        'autodocs',
    ],
    argTypes  : {
        background : {
            description : 'Background styles of the component.',
        },
        maxWidth   : {
            description : 'Maximum width of the component.',
        },
        children   : {
            description : 'Children of the component.',
        },
    },
};

export default meta;

const Label = ({
    children,
} : {
    children : string,
}) => (
    <Typography
        color='grey.400'
        variant='caption'>
        {children}
    </Typography>
);

type Story = StoryObj<typeof Footer>;

export const Default : Story = {
    args : {
        children : [
            <FooterLink key='terms'>
                <Label>Terms and conditions</Label>
            </FooterLink>,
            <FooterLink key='privacy'>
                <Label>Privacy policy</Label>
            </FooterLink>,
            <FooterLink key='disclaimer'>
                <Label>Disclaimer</Label>
            </FooterLink>,
        ],
    },
};
