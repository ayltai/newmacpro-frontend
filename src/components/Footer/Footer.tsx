import { Box, Container, Unstable_Grid2 as Grid, } from '@mui/material';
import React, { FC, ReactElement, } from 'react';

import { default as FooterLink, FooterLinkProps, } from '../FooterLink';
import type { FooterProps, } from './Footer.types';

/**
 * Footers show additional information or navigation at the bottom of a page.
 * @param background The background styles of the component.
 * @param maxWidth The maximum width of the component.
 * @param children The children of the footer, normally FooterLink.
 * @param rest The other props that get passed to the component.
 */
const Footer : FC<FooterProps> = ({
    background = 'black',
    maxWidth   = 'lg',
    children,
    ...rest
}) => {
    const links : ReactElement<FooterLinkProps>[] = [];

    if (children) {
        if (Array.isArray(children)) {
            links.push(...children.filter(child => child.type === FooterLink));
        } else {
            if (children.type === FooterLink) links.push(children);
        }
    }

    return (
        <Box
            sx={{
                background,
            }}
            width='100%'>
            <Container maxWidth={maxWidth}>
                <Grid
                    container
                    justifyContent='space-evenly'
                    textAlign='center'
                    {...rest}>
                    {links.map(link => (
                        <Grid
                            key={link.key}
                            xs={12}
                            sm={6}
                            md={4}
                            lg={3}>
                            {link}
                        </Grid>
                    ))}
                </Grid>
            </Container>
        </Box>
    );
};

export default Footer;
