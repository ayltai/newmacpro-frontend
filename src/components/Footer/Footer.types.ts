import { ReactElement, } from 'react';

import type { FooterLinkProps, } from '../FooterLink';

export interface FooterProps {
    background?      : string,
    maxWidth?        : 'xs' | 'sm' | 'md' | 'lg' | 'xl',
    children?        : ReactElement<FooterLinkProps> | ReactElement<FooterLinkProps>[],
    [ key : string ] : any,
}
