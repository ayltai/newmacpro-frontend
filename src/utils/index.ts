import * as Sentry from '@sentry/react';

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const capitalizeFirstLetter = (input : string) => input.charAt(0).toUpperCase() + input.slice(1);
