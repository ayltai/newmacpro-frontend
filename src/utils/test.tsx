import { ThemeProvider, } from '@mui/material';
import { configureStore, Store, } from '@reduxjs/toolkit';
import { act, render, } from '@testing-library/react';
import mediaQuery from 'css-mediaquery';
import { ConfirmProvider, } from 'material-ui-confirm';
import React, { ReactNode, } from 'react';
import { Provider, } from 'react-redux';

import { bundleApi, cartApi, } from '../apis';
import { POPULAR_APPS, } from '../constants';
import { cartReducer, prefReducer, store as reduxStore, } from '../states';
import { AppTheme, } from '../styles';
import { Cart, } from '../types';

export const mockApis = async (cart? : Cart) => {
    fetchMock.doMock(request => {
        if (request.url.endsWith('/api/bundles') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'GET') return Promise.resolve({
            status : 200,
            body   : JSON.stringify({
                id          : '1',
                displayName : 'Dummy',
                packages    : [
                    POPULAR_APPS[0],
                ],
            }),
        });

        if (request.url.endsWith('/api/bundles/1') && request.method === 'DELETE') return Promise.resolve({
            status : 200,
            body   : 'OK',
        });

        if (request.url.endsWith('/api/carts') && request.method === 'POST') return Promise.resolve({
            status : 201,
            body   : 'Created',
        });

        return Promise.resolve({
            status : 404,
            body   : 'Not found',
        });
    });

    await act(async () => {
        await reduxStore.dispatch(bundleApi.endpoints?.getBundle.initiate({
            bundleId : '1',
            idToken  : 'idToken',
        }));
        if (cart) await reduxStore.dispatch(cartApi.endpoints.addCart.initiate(cart));
    });
};

export const createMatchMedia = (width : number) => (query : string) => ({
    matches             : mediaQuery.match(query, {
        width,
    }),
    media               : '',
    onchange            : null,
    addListener         : () => {},
    removeListener      : () => {},
    addEventListener    : () => {},
    removeEventListener : () => {},
    dispatchEvent       : () => false,
});

export const createStore = (preloadedState? : Record<string, any>) => configureStore({
    preloadedState,
    reducer    : {
        cart                      : cartReducer,
        pref                      : prefReducer,
        [ bundleApi.reducerPath ] : bundleApi.reducer,
        [ cartApi.reducerPath   ] : cartApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(bundleApi.middleware, cartApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <ThemeProvider theme={AppTheme}>
                <ConfirmProvider>
                    {children}
                </ConfirmProvider>
            </ThemeProvider>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
