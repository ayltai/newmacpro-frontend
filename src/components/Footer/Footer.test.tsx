import React from 'react';

import { render, } from '../../utils/test';
import { default as Footer, } from './Footer';

describe('<Footer />', () => {
    it('renders correctly', () => expect(render(<Footer />)).toMatchSnapshot());
});
