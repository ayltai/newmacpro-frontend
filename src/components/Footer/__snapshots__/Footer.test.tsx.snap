// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-qp6edm"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
        >
          <div
            class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-1hrqshc-MuiGrid2-root"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-qp6edm"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-1hrqshc-MuiGrid2-root"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
